embedded/sql/stmt.go

```diff
@@ -695,6 +739,114 @@ func (stmt *UpsertIntoStmt) compileUsing(e *Engine, implicitDB *Database, params
 	return summary, nil
 }
 
+func (e *Engine) fetchPKRow(table *Table, valuesByColID map[uint32]TypedValue) (*Row, error) {
+	pkRanges := make(map[uint32]*typedValueRange, len(table.primaryIndex.cols))
+
+	for _, pkCol := range table.primaryIndex.cols {
+		pkVal := valuesByColID[pkCol.id]
+
+		pkRanges[pkCol.id] = &typedValueRange{
+			lRange: &typedValueSemiRange{val: pkVal, inclusive: true},
+			hRange: &typedValueSemiRange{val: pkVal, inclusive: true},
+		}
+	}
+
+	scanSpecs := &ScanSpecs{
+		index:         table.primaryIndex,
+		rangesByColID: pkRanges,
+	}
+
+	r, err := e.newRawRowReader(e.snapshot, table, 0, table.name, scanSpecs)
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.Close()
+
+	return r.Read()
+}
+
+// deleteIndexEntriesFor mark previous index entries as deleted
+func (e *Engine) deleteIndexEntriesFor(
+	pkEncVals []byte,
+	currValuesByColID, newValuesByColID map[uint32]TypedValue,
+	table *Table,
+	summary *TxSummary) (reusableIndexEntries map[uint32]struct{}, err error) {
+
+	reusableIndexEntries = make(map[uint32]struct{})
+
+	for _, index := range table.indexes {
+		if index.IsPrimary() {
+			continue
+		}
+
+		var prefix string
+		var encodedValues [][]byte
+		var val []byte
+
+		if index.IsUnique() {
+			prefix = UIndexPrefix
+			encodedValues = make([][]byte, 3+len(index.cols))
+			val = append([]byte{1}, pkEncVals...)
+		} else {
+			prefix = SIndexPrefix
+			encodedValues = make([][]byte, 4+len(index.cols))
+			encodedValues[len(encodedValues)-1] = pkEncVals
+			val = []byte{1} // set deletion flag
+		}
+
+		encodedValues[0] = EncodeID(table.db.id)
+		encodedValues[1] = EncodeID(table.id)
+		encodedValues[2] = EncodeID(index.id)
+
+		// some rows might not indexed by every index
+		indexed := true
+
+		// existent index entry is deleted only if it differs from existent one
+		sameIndexKey := true
+
+		for i, col := range index.cols {
+			currVal, notNull := currValuesByColID[col.id]
+			if !notNull {
+				indexed = false
```

Comment:
Is this even possible? Currently we require all indexed columns to have non-null values.